preprocessing-scripts/crop_image_and_detect_aimpoint.py

```diff
@@ -0,0 +1,172 @@
+                return model
+            except Exception as e:
+                logger.error(f"Failed to load the model: {e}")
+                raise
+    else:
+        logger.error(
+            f"Failed to download the checkpoint. Status code: {response.status_code}"
+        )
+        response.raise_for_status()
+
+
+def main(args: argparse.Namespace) -> None:
+    """
+    Process calibration images and extract focal spots using UTIS.
+
+    This function downloads and processes heliostat calibration data before loading the UTIS model from a given URL.
+    It then applies UTIS to detect focal spots for specified heliostats and measurement IDs.
+    """
+    # Ensure the output directory is a Path object.
+    output_dir = Path(args.output_dir)
+
+    # Create STAC client and download heliostat data.
+    client = StacClient(output_dir=output_dir)
+    client.get_heliostat_data(
+        heliostats=args.heliostats,
+        collections=args.collections,
+        filtered_calibration_keys=args.filtered_calibration,
+    )
+
+    # Load the UTIS model directly from the URL.
+    loaded_model = load_model_from_url(mappings.UTIS_MODEL_CHECKPOINT)
+
+    # Process each heliostat and its measurements.
+    for heliostat in args.heliostats:
+        measurement_ids = args.measurement_id
+
+        for measurement_id in measurement_ids:
+            # Define paths for current measurement.
+            image_path = (
+                output_dir
+                / heliostat
+                / mappings.SAVE_CALIBRATION
+                / f"{measurement_id}_raw.png"
+            )
+            calibration_properties_path = (
+                output_dir
+                / heliostat
+                / mappings.SAVE_CALIBRATION
+                / f"{measurement_id}-calibration-properties.json"
+            )
+
+            # Extract target from calibration properties.
+            with open(calibration_properties_path, "r") as file:
+                calibration_data = json.load(file)
+            target = calibration_data[mappings.TARGET_NAME_KEY]
+
+            # Crop the image using template matching.
+            cropped_image = crop_image_with_template_matching(
+                image_path, target, n_grid=256
+            )
+
+            # Convert to grayscale by using only green color channel and convert to torch tensor.
+            cropped_image_tensor = (
+                torch.tensor(cropped_image[:, :, 1], dtype=torch.float32) / 255.0
+            )
+
+            # Detect focal spot.
+            focal_spot = detect_focal_spot(cropped_image_tensor, target, loaded_model)
+
+            # Log the detected aim point.
+            logger.info(
+                "Heliostat %s: For Measurement %s, Focal Spot detected at %s.",
+                heliostat,
+                measurement_id,
+                focal_spot.aim_point.tolist(),
+            )
+
+            # TODO: Write results to CSV or database?
```

Comment:
I think the resulting images should be saved in a directory specified through the arguments, and the saved focal spots should be saved to a single CSV. We don't wan tot save directly to the PAINT database.